src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs

```diff
@@ -1313,6 +1313,93 @@ public async Task DuplexStreaming_RequestCTCancellation_DoesNotApply()
             connection.Dispose();
         }
 
+        [ConditionalTheory(nameof(IsMsQuicSupported))]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task DuplexStreaming_AbortByServer_StreamingCancelled(bool graceful)
+        {
+            if (UseQuicImplementationProvider == QuicImplementationProviders.Mock)
+            {
+                return;
+            }
+
+            var message = new byte[1024];
+            var random = new Random(0);
+            random.NextBytes(message);
+
+            using Http3LoopbackServer server = CreateHttp3LoopbackServer();
+            Http3LoopbackConnection connection = null;
+            Http3LoopbackStream serverStream = null;
+
+            Task serverTask = Task.Run(async () =>
+            {
+                connection = (Http3LoopbackConnection)await server.EstablishGenericConnectionAsync();
+                serverStream = await connection.AcceptRequestStreamAsync();
+
+                HttpRequestData requestData = await serverStream.ReadRequestDataAsync(readBody: false).WaitAsync(TimeSpan.FromSeconds(30));
+
+                // abort the connection, including the just-received connection
+                if (graceful)
+                {
+                    await connection.ShutdownAsync(true);
+                }
+                else
+                {
+                    await connection.CloseAsync(Http3LoopbackConnection.H3_INTERNAL_ERROR);
+                }
+            });
+
+            Task clientTask = Task.Run(async () =>
+            {
+                StreamingHttpContent requestContent = new StreamingHttpContent();
+
+                using HttpClient client = CreateHttpClient();
+                using HttpRequestMessage request = new()
+                {
+                    Method = HttpMethod.Post,
+                    RequestUri = server.Address,
+                    Version = HttpVersion30,
+                    VersionPolicy = HttpVersionPolicy.RequestVersionExact,
+                    Content = requestContent
+                };
+
+                var cts = new CancellationTokenSource();
+
+                var responseTask = client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cts.Token);
+
+                // wait for the content to start transferring
+                Stream requestStream = await requestContent.GetStreamAsync().WaitAsync(TimeSpan.FromSeconds(10)); // the stream is Http3WriteStream
+
+                CancellationToken serializeToken = requestContent.SerializeCancellationToken;
+                Assert.False(serializeToken.IsCancellationRequested);
+
+                // Send headers and first part of the stream
+                await requestStream.FlushAsync();
+
+                await requestStream.WriteAsync(message);
```

Comment:
Now you are only reading headers on the server, but from client you are sending both headers and (separately) some data. That way it is potentially possible to get an exception on the second write, if the server already closed the connection.

I think it's better to either make sure the server consumed the data, i.e. read the same amount of bytes from the content that is sent here (remember it might theoretically arrive as several messages) -- or skip sending the additional data at all, I guess it doesn't really bring much value to the test.

Reply:
Yeah, I think skipping the data is the best option here, since there is no way to drain only part of the data on the server-side (as far as I can see).

Reply:
While there is an (ugly) way to drain only part of the data, see e.g.  https://github.com/dotnet/runtime/blob/719737e5929556eb728de80714ebc02f0aea1cfb/src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs#L1033 I agree there's no need to bother with that 🙂